resolver/src/resolver_handle/mod.rs

```diff
@@ -59,12 +57,82 @@ impl DnsHandle for BasicResolverHandle {
     }
 }
 
-/// A Resolver for DNS records.
-pub struct ResolverFuture {
-    config: ResolverConfig,
-    options: ResolverOpts,
-    pub(crate) client_cache: CachingClient<LookupEither<BasicResolverHandle, StandardConnection>>,
-    hosts: Option<Arc<Hosts>>,
+/// A handle for resolving DNS records.
+///
+/// Creating a `ResolverHandle` returns a new handle and a future that should
+/// be spawned on an executor to drive the background work. The lookup methods
+/// on `ResolverHandle` request lookups from the background task.
+///
+/// The futures returned by a `ResolverHandle` and the corresponding background
+/// task need not be spawned on the same executor, or be in the same thread.
+///  Additionally, one background task may have any number of handles; calling
+/// `clone()` on a handle will create a new handle linked to the same
+/// background task.
+///
+/// *NOTE* If lookup futures returned by a `ResolverHandle` and the background
+/// future are spawned on two separate `CurrentThread` executors, one thread
+/// cannot run both executors simultaneously, so the `run` or `block_on`
+/// functions will cause the thread to deadlock. If both the background work
+/// and the lookup futures are intended to be run on the same thread, they
+/// should be spawned on the same executor.
+///
+/// The background task future will finish once all existing handles to it
+/// have been dropped.
+#[derive(Clone)]
+pub struct ResolverHandle {
```

Comment:
> We should should add an alias at `trust_dns_resolver::ResolverFuture` to this (deprecated) to help people transition to this new name.

I agree. It is worth pointing out, however, that the changed return type from the constructor for this type means that this branch is a breaking change nonetheless, even if we provide an alias...

> Also, what's your current feeling on the `ResolverHandle` name? Like I said in the issue, I'm open to other names... We already have `Resolver` for the synchronous resolver, another option would be:
>
> `TokioResolver`
> `AsyncResolver`
>
> ... others?

Hmm. I'm not particularly attached to `ResolverHandle` either. I like your suggestion of `AsyncResolver` for the following reasons:
+ constructing it with `AsyncResolver::new()` feels a lot better than `ResolverHandle::new`, since we're not _just_ constructing a new handle, 
+ similarly, I think `async_resolver` is a much better name for the module where this type is defined; I don't like calling the module `resolver_handle`, since it contains much more than just the handle type, but the old name `resolver_future` was no longer accurate.
+ I'm inclined to say "no" to `TokioResolver`, since that implies that the resolver is tied directly to `Tokio`, when in actuality, it uses futures in a non-tokio-specific way and could (theoretically) be run on an `Executor` from some library other than `tokio`...

I also wondered about `BackgroundResolver`, but that seems too long IMO.

Reply:
> however, that the changed return type from the constructor for this type means that this branch is a breaking change nonetheless

Great point. Would you mind dropping a breaking change comment (probably needs a new unreleased section) in the resolver/CHANGELOG.md?

> I like your suggestion of AsyncResolver

Great! Let’s do that.